packages/blocks/src/api/parser/get-block-attributes.js

```diff
@@ -295,3 +295,28 @@ export function getBlockAttributes(
 		attributes
 	);
 }
+
+/**
+ * Applies default attributes from a block type to a set of attributes.
+ *
+ * @param {string|Object} blockTypeOrName Block type or name.
+ * @param {?Object}       attributes      Known block attributes (from delimiters).
+ *
+ * @return {Object} Block attributes with defaults applied.
+ */
+export function applyDefaultAttributes( blockTypeOrName, attributes = {} ) {
+	const blockType = normalizeBlockType( blockTypeOrName );
+
+	const blockAttributes = Object.fromEntries(
+		Object.entries( blockType.attributes ?? {} ).map(
+			( [ key, schema ] ) => [
+				key,
+				attributes[ key ] === undefined
+					? schema.default
+					: attributes[ key ],
+			]
+		)
+	);
+
+	return blockAttributes;
+}
```

Comment:
> also I think that it could make sense to move to a more imperative approach in this function, both for efficiency sake (no need to allocate the extra maps/arrays or break shallow-equality for attributes with all the required values) and from a clarity standpoint, to highlight the fill-if-missing-logic

Yeah, it was just a quick copy of `getBlockAttributes` that didn't try to source values from the block content. Looks like everywhere that used to use `_.mapValues` uses the `Object.entries` style since https://github.com/WordPress/gutenberg/pull/50285. It's easy enough to change here, and `getBlockAttributes` should also probably be changed.